### ISSUE TITLE
Adding dynamic tab to an empty panel

### ISSUE THREAD

**Issue by dmenne** [RinteRface/shinyMobile#181]
With default shiny, I can add a dynamically created tab to an empty tabset panel by using target = NULL.

I have not been able to find a equivalent method for shinyMobile; an initial tab is always required. Any workaround?

```
# example app for inserting/removing a tab
ui <- fluidPage(
  sidebarLayout(
    sidebarPanel(
      actionButton("add", "Add 'Dynamic' tab"),
    ),
    mainPanel(
      tabsetPanel(id = "tabs")
    )
  )
)
server <- function(input, output, session) {
  observeEvent(input$add, {
    insertTab(inputId = "tabs",
              tabPanel("Dynamic", "This a dynamically-added tab"),
              target = NULL
    )
  })
}

shinyApp(ui, server)

**Comment by cofrca** [comment]
I'm not sure if this'll work for your application, but for my project I ended up having to use a blank f7Tabs as a placeholder.
To remove the toolbar from the bottom (at the beginning) - I have this blank f7Tabs in html and the toolbar part is intentionally missing `class = "toolbar"` . It's definitely not the cleanest, but it works.

For a better fix I think we'd need to work on the `insert_` function in shinyMobile.js, but I know I'm not super good at js ... Hence my awkward placeholder haha. Below is a condensed version of what I did. I'm still pretty new to this, but I hope this can help in leu of an actual programmatic fix.

```r

library(shiny)
library(shinyMobile)
library(shinyjs)
shinyApp(
  ui = f7Page(
    useShinyjs(),
    title = "Insert a tab Before the target",
    f7TabLayout(
      panels = tagList(
        f7Panel(title = "Left Panel", side = "left", theme = "dark", 
                f7Button(inputId = "gobutton", label = "Add New tabs!"), effect = "cover")
      ),
      navbar = f7Navbar(
        title = "Tabs",
        hairline = FALSE,
        shadow = TRUE,
        leftPanel = TRUE
      ),
      HTML('
     <div id="toolMan" class="tabbar tabbar-labels toolbar-bottom tabLinks">
  <div class="toolbar-inner">
    <a data-tab="#tabs-default" class="tab-link tab-link-active">default</a>
  </div>
</div>
<div class="tabs-swipeable-wrap">
  <div class="tabs ios-edges" id="tabs">
     <div class="page-content tab" data-active="false" id="tabs-default" data-value="Default" data-hidden="false" style="background-color: gainsboro;"></div>
  </div>
</div>')
      #use <div class="tabs-animated-wrap"> if swipeable tabs is not desired
    )
  ),
  server = function(input, output, session) {
    
    
    observeEvent(input$gobutton, {
      addClass(id = "toolMan", class = "toolbar")
      
      insertF7Tab(
        id = "tabs",
        position = "before",
        target = 'default',
        tab = f7Tab(tabName = paste0("newtab", input$gobutton), icon = f7Icon("envelope"), paste0("newtab", input$gobutton)),
        select = TRUE
      )
      insertF7Tab(
        id = "tabs",
        position = "before",
        target = 'default',
        tab = f7Tab(tabName = "AltTab", icon = f7Icon("envelope"), f7Card(h3("This is the alternate tab"))),
        select = FALSE
      )
      #removeUI(selector = '#gobutton')
      removeF7Tab(id = "tabs", target = 'default')
    })
  }
)

```

**Comment by dmenne** [comment]
Assuming that you have a first dummy tab:
```
  f7TabLayout(
    navbar = f7Navbar(title = "Tabs",
                      subtitle = "shinyxlsforms"),
    f7Tabs(
      id = "xlsforms_tabs",
      f7Tab(tabName = "xlsform", active = TRUE, hidden=FALSE)
    )
  )

```
You can remove it with:

```
          removeUI(selector='a[data-tab="#xlsforms_tabs-xlsform"]')
```

I wish there would be a cleaner way to insert a tab into an empty page, but also to hide  a tab after creation.

**Comment by dmenne** [comment]
Another workaround: 

* Create a dummy first tab. If you want to use icons with dynamically created tabs, you must add an icon here. If this is missing, icons are not created and replaced by uppercase names of icons.

```
f7Tab(tabName = "xlsform", icon=f7Icon("envelope"))
```

* Hide the tab with 
`shinyjs::hide(selector='a[data-tab="#xlsforms_tabs-xlsform"]')`

You must initialize the use of shinyjs in the ui with `useShinyjs()`- see shinyjs vignette

Now you can add items to the hidden tab.  f7Tab has a hidden=TRUE parameter, but the trick does not work when you hide the dummy with this parameter. I believe this is a bug, because the parameter hidden was added for a similar use.

**Comment by cofrca** [comment]
Good thinking, It is funny how we came to the same basic idea.

I really like the idea of using `shinyjs::hide()`  - I used this to hide the entire bottom toolbar:
 `shinyjs::hide(selector='div[class="toolbar tabbar tabbar-labels toolbar-bottom tabLinks"]')`

^This let me move away from the `HTML()` version of the dummy tab that I included in my first example. My use of `shinyjs::addClass()` _worked_, but was definitely not as elegant.


Correct me if I'm wrong, but I don't think that f7Tab's hidden=TRUE parameter is a bug. If I recall correctly, the hidden=TRUE creates the tab normally but does not add a link on the toolbar (leading to a tab that you can only reach via `updateF7Tabs()` )
Whereas `shinyjs::hide()` hides the selected element from view

**Comment by dmenne** [comment]
You are right, it is probably by design (see https://github.com/RinteRface/shinyMobile/issues/44) but somewhat confusingly named.

**Comment by cofrca** [comment]
I was looking into `insertF7Tab()` and the `insert_` function in shinyMobile.js, My thought was to try to edit these functions to add the functionality we were looking for.

The R function  `insertF7Tab()` has arguments for the **target** tab, as well as the **position** (before/after) that target. Then the javascript side of things uses the `.insertBefore()` and `.insertAfter()` methods to insert the new tab and tab icon into the page in **position** with the **target** tab.  
My thought was to add a third **position**, let's call it 'new', and use the `.appendTo()` method to add the new 'page-content tab' to the end of the f7tabs. I was able to apply this with mild success - the new page was inserted at the end with no trouble and the icons / swipe controls worked as expected. 
My next step was to start with an empty `f7Tabs()`, and use this new technique to add a dynamically created `f7Tab` to an empty `f7Tabs` page. However, when I tried this approach all tab content did not display properly and instead merged together (the toolbar still looked normal)

![image](https://user-images.githubusercontent.com/79738584/114494146-d4836c00-9be9-11eb-8510-4da2ea384b72.png)


I might be overlooking something very simple, I haven't used js much, so I am likely to overlook things due to my unfamiliarity. These are the changes I made   [cofrca-#181](https://github.com/RinteRface/shinyMobile/compare/master...cofrca:cofrca-%23181)  (also new to GitHub, is this the best way to share changes in progress?)

<hr>

My test code is:

```r
library(shiny)
library(shinyMobile)
shinyApp(
  ui = f7Page(
    title = "Insert a tab Before the target",

    f7TabLayout(
      panels = tagList(
        f7Panel(title = "Left Panel", side = "left", theme = "light", "Blabla",
                                   f7Button(inputId = "go", label = "Go"), effect = "cover"),
        f7Panel(title = "Right Panel", side = "right", theme = "dark", "Blabla", effect = "cover")
      ),
      navbar = f7Navbar(
        title = "Tabs",
        hairline = FALSE,
        shadow = TRUE,
        leftPanel = TRUE,
        rightPanel = TRUE
      ),
      f7Tabs(
        animated = FALSE,
        swipeable = TRUE,
        id = "tabs"
      )
    )
  ),
  server = function(input, output, session) {
    observeEvent(input$go, {

      if(input$go == 1){
        insertF7Tab(
          id = "tabs",
          position = "new",
          tab = f7Tab (tabName = "initialTab",icon = f7Icon("smiley"), "initialTab Test"),
          select = F
        )
      }
      if(input$go == 2){
        insertF7Tab(
          id = "tabs",
          position = "after",
          target = "initialTab",
          tab = f7Tab(
            tabName = "Tab 2",
            icon = f7Icon("today"),
            active = FALSE,
            "Tab 2"
          ),
          select = F
        )
        insertF7Tab(
          id = "tabs",
          position = "after",
          target = "initialTab",
          tab = f7Tab(
            tabName = "Tab 1",
            icon = f7Icon("envelope"),
            active = TRUE,
            "Tab 1"

          ),
          select = F
        )
      }

    })
  }
)

```

**Comment by cofrca** [comment]
Follow-up:

I got it working, but it requires an existing **blank** `f7Tabs()` in the UI.  If the position 'new' is used when there is an existing tab in place, there are sometimes odd effects.  Even when working properly, This method still leaves a blank toolbar at the bottom of the page.
Because of these restrictions, I believe the existing workaround provides more flexibility and is a better solution for our niche use.

**Comment by DivadNojnarg** [comment]
Finally fixed in `rc-1.0.0`.
